### PR TITLE
Fix for #516 - Slashes in BkBasic

### PR DESCRIPTION
**Pull request by alexey-krylov**
Fix for #516.
Added slashes fix + junit test for this check.
Maybe it will be nice to expose header name for future reuse.

### CODE REVIEW

**Comment by davvd** [comment]
@alexey-krylov Many thanks, I will find someone to review it before we merge

**Comment by davvd** [comment]
@dmzaytsev review this one, please

**Comment by dmzaytsev** [comment]
@alexey-krylov you have checkstyle issues, see AppVeyor's or Travis's log
let's fix it

**Comment by dmzaytsev** [comment]
@alexey-krylov according the quality rules http://at.teamed.io/qa.html
>Pull request description explains the solution proposed and contains a link to the original ticket it is related to.

please update the pull request description

**Comment by alexey-krylov** [comment]
Resolving #516.
Code reformatted according to teamed.io rules.

**Comment by alexey-krylov** [comment]
@dmzaytsev please check - codestyle is fixed now. I have to suppress ExcessiveImports on BkBasicTest because it is initially overloaded. Also found that PMD suppression is a usual practice in 'takes' junits.
As an option - move headers check to separate test case.

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 1]
```diff
@@ -51,14 +61,15 @@
...
  * @author Dmitry Zaytsev (dmitry.zaytsev@gmail.com)
  * @version $Id$
- * @since 0.15.2
  * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  * @checkstyle MultipleStringLiteralsCheck (500 lines)
  * @todo #306:30min At the moment we don't support HTTP
- *  persistent connections. Would be great to implement
- *  this feature. BkBasic.accept should handle more
- *  than one HTTP request in one connection.
+ * persistent connections. Would be great to implement
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov Travis build failed because of invalid format in the puzzle. You have to add a space from second line

>/home/travis/.rvm/gems/ruby-1.9.3-p551/gems/pdd-0.15.4/lib/pdd/source.rb:144:in `rescue in puzzles': ["in /home/travis/build/yegor256/takes/src/test/java/org/takes/http/BkBasicTest.java", #<PDD::Error: Space expected>] (PDD::Error)

**Comment by dmzaytsev** [comment]
@alexey-krylov the original ticket mentioned `UnknownHostException` on `getLocalAddress` and `getRemoteAddress` calls
However I hadn't saw these calls in your test
You have to reproduce the issue in a test first
and then fix it

**Comment by dmzaytsev** [comment]
@alexey-krylov 
>Messages in a ticket always start with a name of a user they are addressed to.
see http://at.teamed.io/qa.html

please fix your comment

> alexey-krylov commented 8 hours ago
> Resolving #516.
> Code reformatted according to teamed.io rules.

**Comment by alexey-krylov** [comment]
@dmzaytsev is it possible to find puzzle invalid format errors before commit to Travis? I run 'mvn clean install -Pqulice' and it works fine for code that i modified.

**Comment by dmzaytsev** [comment]
@alexey-krylov yep, this [line](https://github.com/yegor256/takes/blob/master/.travis.yml#L10) checks for correct puzzle
see also https://github.com/teamed/pdd how to install pdd locally

**Comment by alexey-krylov** [comment]
@dmzaytsev all checks have passed :) Please check.
Also i added address-related check.

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 2]
```diff
@@ -50,15 +61,20 @@
...
  * @author Dmitry Zaytsev (dmitry.zaytsev@gmail.com)
  * @version $Id$
- * @since 0.15.2
  * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  * @checkstyle MultipleStringLiteralsCheck (500 lines)
  * @todo #306:30min At the moment we don't support HTTP
  *  persistent connections. Would be great to implement
  *  this feature. BkBasic.accept should handle more
  *  than one HTTP request in one connection.
+ * @since 0.15.2
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov why did you move this line?

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 3]
```diff
@@ -135,7 +134,71 @@ public void exec(final URI home) throws IOException {
     }
 
     /**
+     * BkBasic produces headers with addresses without slashes.
+     *
+     * @throws IOException If some problem inside
+     */
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov let's add a new puzzle to refactoring this test to use real sockets instead mocked object
see https://github.com/yegor256/takes/pull/426

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 4]
```diff
@@ -135,7 +134,71 @@ public void exec(final URI home) throws IOException {
...
+    @Test
+    public void addressesInHeadersAddedWithoutSlashes() throws IOException {
+        final Socket socket = BkBasicTest.createMockSocket();
+        final AtomicReference<Request> ref = new AtomicReference<Request>();
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov why do we need `AtomicReference` here?
`BkBasic` doesn't use threads as I know

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 5]
```diff
@@ -135,7 +134,71 @@ public void exec(final URI home) throws IOException {
...
+        );
+        for (final String header : found) {
+            MatcherAssert.assertThat(
+                header,
+                Matchers.not(
+                    Matchers.containsString("/")
+                )
+            );
+        }
+        MatcherAssert.assertThat(
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov I think this matcher and below will be enough for the test
we have a ticket 'there is an exception on get...()', so you reproduced the problem and fix it

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 6]
```diff
@@ -182,7 +245,7 @@ public void run() {
                 // @checkstyle MagicNumber (1 line)
                 final byte[] buffer = new byte[4096];
                 for (int count = input.read(buffer); count != -1;
-                    count = input.read(buffer)) {
+                     count = input.read(buffer)) {
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov please remove the space

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 7]
```diff
@@ -239,7 +303,7 @@ public void run() {
...
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov and here

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 8]
```diff
@@ -297,7 +362,7 @@ public void run() {
...
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: and here

**Comment by dmzaytsev** [comment]
@alexey-krylov please see a few comments above

**Comment by alexey-krylov** [comment]
@dmzaytsev this spaces added by supposed codestyle from @yegor256 and caused by limited to 80 chars maximum line lenght (because 'for' statement is too long for 80 chars). Also this style did not cause warnings in Qulice.
I dont think that manual code reformatting is a good and productive way to create clean code.
I will be very appreciate to you if your share IDE config with this check.

Regarding to another comments i will fix them.

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 2 above)
Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev I don't see any repeatable correlation of this statement position. For example in BkParallelTest @since placed last, after @checkstyle annotation
```java
/**
 * Test case for {@link BkParallel}.
 *
 * @author Dmitry Zaytsev (dmitry.zaytsev@gmail.com)
 * @version $Id: da765ae26ed32c19b2361ee48ab723140a1cfa93 $
 * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
 * @since 0.15.2
 */
```

But in FtCLITest placing is changed:
```java
/**
 * Test case for {@link FtCLI}.
 * @author Yegor Bugayenko (yegor@teamed.io)
 * @version $Id: 6c8a3e746b7f0ba3b9010afac6fd3c3403a34036 $
 * @since 0.1
 * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
 */
```

So is there any standartized check for this?

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 3 above)
Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev you mean to add @todo PDD block?

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 8 above)
Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev this spaces added by supposed codestyle config for IDEA from @yegor256 and caused by limited to 80 chars maximum line lenght (because 'for' statement is too long for 80 chars). Also this style did not cause warnings in Qulice.
I dont think that manual code reformatting is a good and productive way to create clean code.
I will be very appreciate to you if your share IDE config with this check.

Regarding to another comments i will fix them.

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 5 above)
Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev I agree, but headers will be set only in BkBasic.accept. Method BkBasic.addSocketHeaders has private static final, so it is non testable without reflection-based hacks with changing access modifiers. So to create RqWithHeaders i need to 'catch' it first.
I do that in upper code.

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 3 above)
Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev it will be more correct to define @todo at class level, because createMockSocket used twice. Not only in my test

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 2 above)
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov this change not related to the ticket that's why I asked about it
about tag orders you can read here [Order of Tags](http://www.oracle.com/technetwork/articles/java/index-137868.html)

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 3 above)
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov yes, that's exactly what I mean.

**Comment by dmzaytsev** [comment]
@alexey-krylov okay let's keep it as is
I use same config as Yegor
and usually the indentation step is 4 spaces

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 4 above)
Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev replaced to holder array.

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 9]
```diff
@@ -135,7 +135,71 @@ public void exec(final URI home) throws IOException {
...
 
     /**
+     * BkBasic produces headers with addresses without slashes.
+     *
+     * @throws IOException If some problem inside
+     */
+    @Test
+    public void addressesInHeadersAddedWithoutSlashes() throws IOException {
+        final Socket socket = BkBasicTest.createMockSocket();
+        final Request[] holder = new Request[1];
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov I think it can be like this
```
final Request request;
....
         public Response act(final Request req) {
              request = req;
                  return new RsEmpty();
....
).accept(socket);
```

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 5 above)
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov  you can extract a header from the request like this
```
  new RqHeaders.Smart(new RqHeaders.Base(request)).single(<header name>)
```

or with default value
```
  new RqHeaders.Smart(new RqHeaders.Base(request)).single(<header name>, <default>)
```

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 9 above)
Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev i think only Chuck Norris can assign to final variable :)

Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov right :) my bad, let's revert back to AtomicReference

**Comment by alexey-krylov** [comment]
@dmzaytsev test rewritten by your comments. pls check

**Comment by alexey-krylov** [comment]
@dmzaytsev done

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 6 above)
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov the indentation step should be four spaces. Qulice check such cases from `0.15`
see https://github.com/teamed/qulice/issues/527
please remove the space

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 10]
```diff
@@ -239,7 +291,7 @@ public void run() {
                 // @checkstyle MagicNumber (1 line)
                 final byte[] buffer = new byte[4096];
                 for (int count = input.read(buffer); count != -1;
-                    count = input.read(buffer)) {
+                     count = input.read(buffer)) {
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov please remove the space

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 11]
```diff
@@ -297,7 +350,7 @@ public void run() {
...
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov please remove the space

**Comment by dmzaytsev** [comment]
@alexey-krylov please see a few cosmetic commantrs above and we are ready to merge

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 11 above)
Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev you said earlier that 'okay let's keep it as is'. 
When you do reformat code with IDEA+Yegor profile it places this line without this space?

Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev i offer to exchange with settings via https://gitlab.i-free.com/mretail/idea-settings.git plugin. This is default option of IDEA. We can create teamed.io special repo for configs. This will significantly reduce codestyle frustration.

Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov 
> @dmzaytsev you said earlier that 'okay let's keep it as is'. 

I'm sure Yegor will ask you remove the spaces. I just try to save your time

> When you do reformat code with IDEA+Yegor profile it places this line without this space?

No I didn't pull your branch,  I see only the diff

Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @alexey-krylov I would suggest submit a ticket https://github.com/teamed/xdsd

>  i offer to exchange with settings via https://gitlab.i-free.com/mretail/idea-settings.git plugin. This is default option of IDEA. We can create teamed.io special repo for configs. This will significantly reduce codestyle frustration.

**Comment by alexey-krylov** [comment]
@dmzaytsev done with spaces.

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 11 above)
Comment (alexey-krylov) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev Thanks for link.
Done with issue: https://github.com/teamed/xdsd/issues/27

**Comment by dmzaytsev** [comment]
@alexey-krylov thank you!

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@ypshenychka please, check this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by ypshenychka** [comment]
@dmzaytsev 
>dmzaytsev added a note 2 days ago
and here

According to our [QA Rules](http://at.teamed.io/qa.html):
>Messages in a ticket always start with a name of a user they are addressed to.

Please correct your code review comment by indicating an addressee in the beginning.

**Comment by dmzaytsev** [comment]
@ypshenychka fixed, thanks

**Comment by ypshenychka** [comment]
@dmzaytsev Thank you!

**Comment by ypshenychka** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@ypshenychka thanks for the QA review, we'll work better next time

**Comment by davvd** [comment]
@dmzaytsev just added 10 mins to @ypshenychka (for QA), payment ID is `75304095`

Thank you, I have added **44 mins** to your account in payment/transaction "`75304098`", time consumed: 52 hours and 6 mins

review comments (c=29) added as a bonus

added +44 to your rating, now it is equal to [+2486](http://www.netbout.com/b/35166?open=rating)
